pkg/operator/observe_config.go

```diff
@@ -90,69 +89,114 @@ func (c ConfigObserver) sync() error {
+			Message: strings.Join(messages, "\n"),
+		})
+	} else {
+		condition := v1alpha1helpers.FindOperatorCondition(status.Conditions, v1alpha1.OperatorStatusTypeFailing)
+		if condition != nil && condition.Status != v1alpha1.ConditionFalse && condition.Reason == configObservationErrorConditionReason {
+			condition.Status = v1alpha1.ConditionFalse
+			condition.Reason = ""
+			condition.Message = ""
+		}
+	}
+
+	if !equality.Semantic.DeepEqual(operatorConfig.Status, status) {
+		operatorConfig.Status = *status
+		_, err = c.operatorConfigClient.KubeControllerManagerOperatorConfigs().UpdateStatus(operatorConfig)
+		if err != nil {
+			return err
+		}
 	}
 
 	return nil
 }
 
-// observeClusterConfig observes cloud provider configuration from
-// cluster-config-v1 in order to configure kube-controller-manager's cloud
+// observeCloudProviderNames observes cloud provider configuration from
+// cluster-config-corelistersv1 in order to configure kube-controller-manager's cloud
 // provider.
-func observeClusterConfig(kubeClient kubernetes.Interface, clientConfig *rest.Config, observedConfig map[string]interface{}) (map[string]interface{}, error) {
-	clusterConfig, err := kubeClient.CoreV1().ConfigMaps("kube-system").Get("cluster-config-v1", metav1.GetOptions{})
+func observeCloudProviderNames(listers Listers, currentConfig map[string]interface{}) (observedConfig map[string]interface{}, errs []error) {
+	observedConfig = map[string]interface{}{}
+	cloudProvidersPath := []string{"extendedArguments", "cloud-provider"}
+	currentCloudProvider, _, _ := unstructured.NestedStringSlice(currentConfig, cloudProvidersPath...)
+	clusterConfig, err := listers.configmapLister.ConfigMaps("kube-system").Get("cluster-config-v1")
 	if errors.IsNotFound(err) {
-		glog.Warningf("cluster-config-v1 not found in the kube-system namespace")
-		return observedConfig, nil
+		glog.Warning("configmap/cluster-config-v1.kube-system: not found")
+		return
 	}
 	if err != nil {
-		return observedConfig, err
+		errs = append(errs, err)
+		if len(currentCloudProvider) > 0 {
+			unstructured.SetNestedStringSlice(observedConfig, currentCloudProvider, cloudProvidersPath...)
+		}
+		return
 	}
 
 	installConfigYaml, ok := clusterConfig.Data["install-config"]
 	if !ok {
-		return observedConfig, nil
+		glog.Warning("configmap/cluster-config-v1.kube-system: install-config not found")
+		return
 	}
 	installConfig := map[string]interface{}{}
 	err = yaml.Unmarshal([]byte(installConfigYaml), &installConfig)
 	if err != nil {
 		glog.Warningf("Unable to parse install-config: %s", err)
-		return observedConfig, nil
+		return
 	}
 
 	// extract needed values
 	//  data:
 	//   install-config:
 	//     platform:
 	//       aws: {}
-	platform, ok := installConfig["platform"].(map[string]interface{})
-	if !ok {
-		glog.Warningf("Unable to parse install-config: %s", err)
-		return observedConfig, nil
-	}
-
-	cloudProvider := ""
-	switch {
-	case platform["aws"] != nil:
```

Comment:
looks like you've lost this and made it much harder to extend to our other cloud providers.  please restore.